design/lakectl-testing.md

```diff
@@ -0,0 +1,102 @@
+
+#### **Live Server**
+
+On the upside:
+* Straightforward - the implementation is already there, just bring the server up.
+  * Moreover, `lakefs` modifications are immediately reflected into the tests
+* The best possible simulation - this is the real server `lakectl` is communicating with
+
+And then, on the downside:
+* Bringing up the server requires additional dependencies - underlying storage, DB etc. While this supports the advantage of testing a **real** server environment, it complicates the test setup, and although this is easily solved by containers, it still worth mentioning
+* Performance-wise, creating a server environment for each test (thousands of them, ideally) is very expensive. We can, of course, group several tests (all of them?) to use a single server session, but this will eliminate the isolative and stateless nature of the tests, which is undesired
+* Flexibility is limited to `lakefs` server implementation. This means that testing `lakectl` behavior, with improbable response from `lakefs` server (e.g. timeout, disconnection etc.) becomes very difficult
+
+#### Mock
+
+Advantages:
+* Totally independent from the environment. No need for DB, underlying storage, or any other future dependency that may be added to `lakefs` server. Moreover, future removal of any dependency from `lakefs` (Postgres?!) will not require framework modification. And, once again, although this advantage is pretty much diminished by the usage of containers, it is still worth noting
+* No uptime, as the mock is a part of the test itself
+* Totally stateless - again, as the mock is part of each test
+* Maximal flexibility - each desired server response can be easily hard-coded
+* No need for initial state setup steps
+* Server bugs are not included
+
+Disadvantages:
+* Requires implementation - unlike `lakefs` server, we currently don't have a mock
+* Requires (a lot of) maintenance - every new feature and every bug fix in `lakefs` may, and probably will, need to be reflected in the mock
+
+#### Conclusion ?
+
+For the long term it look like the better approach is to use a full scale mock, which should evolve as tests are created. While this will certainly have an impact on new tests creation ease, the benefits of lack independency on `lakefs` are too important to ignore, especially when dealing with a young product such as `lakefs` which may be subdued to significant changes in the future
+
+### How should we test `lakectl`
+
+While the decision on `lakefs` usage above, has many significant aspects, the 2 approaches on how to test `lakectl` - as a binary or as code - seems to have much lesser impact. Here are some of the advantages and disadvantages of both
+
+#### Testing `lakectl` Binary
+
+Pros:
+* Tests are performed on the real binary and includes all the surroundings, such as env variables
+* Allows usage of 3rd party testing tools, that were designed to test CLIs
+
+Cons:
+* May be hard to debug a failed test
```

Comment:
_Commented by @nopcoder_
Depends - sometimes it is very easy that you can your command line that didn't print out the expected result. You just need to execute it again to debug what failed.

Reply:
Right. I will emphasize the "May" part